network/crl.go

```diff
@@ -0,0 +1,370 @@
+		return nil, err
+	}
+
+	if crl.TBSCertList.NextUpdate.Before(time.Now()) {
+		log.Warnf("CRL: CRL at %s is outdated", uri)
+		return nil, ErrOutdatedCRL
+	}
+
+	if v.Config.cacheTime > 0 {
+		cacheItem := &CRLCacheItem{Fetched: time.Now(), CRL: *crl}
+		v.Cache.Put(uri, cacheItem)
+	}
+
+	return crl, nil
+}
+
+func (v DefaultCRLVerifier) verifyCertWithIssuer(cert, issuer *x509.Certificate) error {
+	log.Debugf("CRL: Verifying '%s'", cert.Subject.String())
+
+	if len(v.Config.uri) == 0 && v.Config.fromCert == crlFromCertIgnore {
+		log.Debugln("CRL: Skipping check since no config URI specified and we were told to ignore URIs from certificate")
+		return nil
+	}
+
+	for _, crlDistributionPoint := range cert.CRLDistributionPoints {
+		log.Debugf("CRL: certificate contains CRL URI: %s", crlDistributionPoint)
+	}
+
+	queriedCRLs := make(map[string]struct{})
+
+	if v.Config.uri != "" {
+		crl, err := v.getCachedOrFetch(v.Config.uri, issuer)
+		if err != nil {
+			log.WithError(err).Debugf("CRL: Cannot get CRL from '%s'", v.Config.uri)
+			return err
+		}
+
+		for _, revokedCertificate := range crl.TBSCertList.RevokedCertificates {
+			if revokedCertificate.SerialNumber.Cmp(cert.SerialNumber) == 0 {
+				log.Warnf("CRL: Certificate %v was revoked at %v", cert.SerialNumber, revokedCertificate.RevocationTime)
+				return ErrCertWasRevoked
+			}
+		}
+
+		queriedCRLs[v.Config.uri] = struct{}{}
+		log.Debugln("CRL: OK, not found in list of revoked certificates")
+	}
+
+	if len(cert.CRLDistributionPoints) > 0 && v.Config.fromCert != crlFromCertIgnore {
+		for _, crlDistributionPoint := range cert.CRLDistributionPoints {
+			if _, ok := queriedCRLs[crlDistributionPoint]; ok {
+				continue
+			}
+
+			log.Debugf("CRL: using '%s' from certificate", crlDistributionPoint)
+
+			crl, err := v.getCachedOrFetch(crlDistributionPoint, issuer)
+			if err != nil {
+				log.WithError(err).Debugf("CRL: Cannot get CRL from '%s'", crlDistributionPoint)
+				return err
+			}
+
+			for _, revokedCertificate := range crl.TBSCertList.RevokedCertificates {
+				if revokedCertificate.SerialNumber.Cmp(cert.SerialNumber) == 0 {
+					log.Warnf("CRL: Certificate %v was revoked at %v", cert.SerialNumber, revokedCertificate.RevocationTime)
+					return ErrCertWasRevoked
+				}
+			}
+
+			queriedCRLs[v.Config.uri] = struct{}{}
+			log.Debugln("CRL: OK, not found in list of revoked certificates")
+		}
+	}
+
+	return nil
+}
+
+// Verify ensures configured CRLs do not contain certificate from passed chain
+func (v DefaultCRLVerifier) Verify(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	for _, chain := range verifiedChains {
+		if len(chain) == 1 {
+			// This one cert[0] must be trusted since it was allowed by more basic verifying routines.
+			// If we are at this point, we have nothing to do, and no CA means no OCSP.
+			log.Debugln("OCSP: Certificate chain contains one certificate, nothing to do")
+			return nil
+		}
```

Comment:
I guess we better invite @storojs72 to discuss this question.
My thought was that if there is only cert in chain, it must be in list of trusted certificates (like the list where root certs are stored)
```go
	// NewTLSConfig() in network/tls_wrapper.go
	return &tls.Config{
		RootCAs:               roots,
		ClientCAs:             roots,
```
otherwise it will be rejected even before OCSP/CRL verification is performed. See [TLS config docs](https://pkg.go.dev/crypto/tls#Config)
```
	// VerifyPeerCertificate, if not nil, is called after normal
	// certificate verification by either a TLS client or server. It
	// receives the raw ASN.1 certificates provided by the peer and also
	// any verified chains that normal processing found. (...)

	// If normal verification fails then the handshake will abort before
	// considering this callback. (...)
```
Simply rejecting single-cert-chains may also be a solution because why would one use such configuration for database encryption?

Reply:
> My thought was that if there is only cert in chain, it must be in list of trusted certificates (like the list where root certs are stored) ... otherwise it will be rejected even before OCSP/CRL verification is performed.
Generally I agree with such approach. But, it seems to me, it's worth considering the situation when the same certificate is also provided in `--tls_ca` argument. In such case, I think, we should also trust the transmitted certificate.

Reply:
> certificate is also provided in `--tls_ca`

But can we use CA certificate for data encryption? They are often (always?) made only to sign other certificates?
Or maybe that "basic TLS verification" will reject such certificates before attempting to check OCSP/CRL?